pkg/engine/internal/scheduler/scheduler.go

```diff
@@ -532,23 +569,57 @@ func (s *Scheduler) RemoveStreams(ctx context.Context, streams ...*workflow.Stre
 
 	var errs []error
 
-	for _, streamToRemove := range streams {
-		registered, exist := s.streams[streamToRemove.ULID]
-		if !exist {
-			errs = append(errs, fmt.Errorf("stream %s not registered", streamToRemove.ULID))
-			continue
+	// First, validate the manifest and ensure that all streams and tasks are recognized.
+	for _, stream := range manifest.Streams {
+		if _, exist := s.streams[stream.ULID]; !exist {
+			errs = append(errs, fmt.Errorf("manifest contains unregistered stream %s", stream.ULID))
 		}
+	}
+	for _, task := range manifest.Tasks {
+		if _, exist := s.tasks[task.ULID]; !exist {
+			errs = append(errs, fmt.Errorf("manifest contains unregistered task %s", task.ULID))
+		}
+	}
 
-		// Check whether the stream has non-terminal tasks associated with it.
-		// In this case, we don't care if there's a local receiver listening.
-		if receiver, ok := s.tasks[registered.taskReceiver]; ok && !receiver.status.State.Terminal() {
-			errs = append(errs, fmt.Errorf("stream %s has active tasks", streamToRemove.ULID))
-			continue
-		} else if sender, ok := s.tasks[registered.taskSender]; ok && !sender.status.State.Terminal() {
-			errs = append(errs, fmt.Errorf("stream %s has active tasks", streamToRemove.ULID))
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	// Remove tasks first. If any of them are currently running, we'll cancel them.
+	for _, taskToRemove := range manifest.Tasks {
+		registered := s.tasks[taskToRemove.ULID] // Validated to exist above
+
+		// Immediately clean up our own resources.
+		s.deleteTask(registered)
+
+		if changed, _ := registered.setState(workflow.TaskStatus{State: workflow.TaskStateCancelled}); !changed {
```

Comment:
Not for this PR, but I wonder if we should reserve Cancelled for an error condition (like context cancelled) and use a non-error terminal state here. 

I'm wondering if we tracked the total number of canceled tasks if it would be noisy and hard to extract insight from due to all the ways a task could end up in the canceled state (vs completed or failed)